pkg/operators/common/usergroupcache.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/s3rj1k/go-fanotify/fanotify"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/host"
+)
+
+// UserGroupCache is a cache of user names, uids, group names and gids
+type UserGroupCache interface {
+	Start()
+	Stop()
+
+	GetUsername(uint32) string
+	GetGroupname(uint32) string
+}
+
+type userGroupEntry struct {
+	id   uint32
+	name string
+}
+
+type userGroupCache struct {
+	notify *fanotify.NotifyFD
+
+	userCache  *ResourceCache[userGroupEntry, uint32]
+	groupCache *ResourceCache[userGroupEntry, uint32]
+
+	loopFinished  chan struct{}
+	useCount      int
+	useCountMutex sync.Mutex
+}
+
+var (
+	userGroupCacheSingleton *userGroupCache
+	userGroupErr            error
+	userGroupOnce           sync.Once
+)
+
+func GetUserGroupCache() (UserGroupCache, error) {
+	userGroupOnce.Do(func() {
+		userGroupCacheSingleton, userGroupErr = newUserGroupCache()
+	})
+	return userGroupCacheSingleton, userGroupErr
+}
+
+func newUserGroupCache() (*userGroupCache, error) {
+	toKey := func(entry *userGroupEntry) (uint32, error) {
+		return entry.id, nil
+	}
+
+	return &userGroupCache{
+		userCache:  NewResourceCache[userGroupEntry](toKey),
+		groupCache: NewResourceCache[userGroupEntry](toKey),
+	}, nil
+}
+
+func (cache *userGroupCache) Start() {
+	cache.useCountMutex.Lock()
+	defer cache.useCountMutex.Unlock()
+
+	// No uses before us, we are the first one
+	if cache.useCount == 0 {
+		fanotifyFlags := uint(unix.FAN_CLOEXEC | unix.FAN_NONBLOCK)
+		openFlags := os.O_RDONLY | unix.O_LARGEFILE | unix.O_CLOEXEC
+		notify, err := fanotify.Initialize(fanotifyFlags, openFlags)
+		if err != nil {
+			log.Warnf("UserGroupCache: initialize fanotify: %v", err)
+			return
+		}
+		defer func() {
+			// Only close the file if we are not going to use it
+			if notify != nil {
+				notify.File.Close()
+			}
+		}()
+
+		passwdPath := filepath.Join(host.HostRoot, "/etc/passwd")
+		err = notify.Mark(unix.FAN_MARK_ADD, unix.FAN_MODIFY, -1, passwdPath)
```

Comment:
Do you get the `FAN_MODIFY` event correctly after several changes? Does `notify.Mark` monitor the inode or the path? I noticed that `useradd` writes a new passwd file and atomically rename it to the correct path.
```
$ sudo strace -f useradd janetest
...
openat(AT_FDCWD, "/etc/passwd+", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 5
umask(022)                              = 0777
fchown(5, 0, 0)                         = 0
fchmod(5, 0644)                         = 0
newfstatat(5, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
write(5, "root:x:0:0:root:/root:/bin/bash\n"..., 3705) = 3705
fsync(5)                                = 0
close(5)                                = 0
newfstatat(AT_FDCWD, "/etc/passwd", {st_mode=S_IFREG|0644, st_size=3656, ...}, AT_SYMLINK_NOFOLLOW) = 0
rename("/etc/passwd+", "/etc/passwd")   = 0
...
```

Reply:
Doesn't seem to. I interpreted the documentation wrong
>  If pathname is absolute, it defines the filesystem object to be marked, and dirfd is ignored.


So I would need to mark `/etc` i guess